scipy/stats/_distn_infrastructure.py

```diff
@@ -2230,6 +2279,33 @@ def _nnlf_and_penalty(self, x, args):
             return -np.sum(logpdf[finite_logpdf], axis=0) + penalty
         return -np.sum(logpdf, axis=0)
 
+    def _censored_nnlf_and_penalty(self, x, args):
```

Comment:
This looks right, but I want to consider a refactoring. I think this can be fit pretty naturally into _nnlf_and_penalty because the penalties are assessed pretty much the same way; the only real difference is how the NLF is calculated. If that part were broken out into separate functions, the rest of the code could be shared. 

Alternatively, the names of `_penalized_nnlf` and `_nnlf_and_penalty` don't really explain why they should be separate functions, and it's always seemed strange that `_nnlf_and_penalty` does both `_nnlf` and `_penalty` together. Maybe keep `_penalized_nnlf` and break `_penalty` and `_nnlf` into separate functions.

Anyway, I guess the thought is that while we're digging in here we might as well clean it up a bit. If you don't want to do it, I suppose i can submit a PR to your branch with what I have in mind.

Reply:
I pushed a bunch of changes yesterday, including a rewrite of `_nnlf_and_penalty`.  It now handles both a regular array and a CensoredData instance, so `_censored_nnlf_and_penalty` is gone.

As it is now, `_penalized_nnlf` takes care of the `loc/scale` transformation.  It then passes the transformed data to `_nnlf_and_penalty` to compute the NLF for the given shape parameters.  The penalty is computed as part of the evaluation of the NLF--data values outside the support and data values where the contribution to the NLF is infinite (presumably because the computed PDF or probability underflowed to 0) are penalized.